File: frontend/src/shared/toolIdentity.ts
import toolIdentity from './toolIdentity.json'

export interface RuntimeToolIdentity {
  label: string
  family: string
  summary: string
  icon: string
}

export interface ResolvedToolIdentity extends RuntimeToolIdentity {
  name: string
  rawName: string
  known: boolean
}

const TOOL_IDENTITIES: Record<string, RuntimeToolIdentity> = toolIdentity
const EVIDENCE_RECORD_LABELS: Record<string, string> = {
  dataset_profile: '记录数据画像',
  verification_fact: '记录验证事实',
  parameter_snapshot: '记录参数快照',
  claim_trace: '记录声明追踪',
}

export function canonicalToolName(toolName: string | null | undefined): string {
  let normalized = (toolName ?? '').trim()
  if (normalized.startsWith('mcp__geospatial__')) {
    normalized = normalized.slice('mcp__geospatial__'.length)
  }

  while (normalized.startsWith('geospatial_geospatial_')) {
    normalized = normalized.slice('geospatial_'.length)
  }

  if (TOOL_IDENTITIES[normalized]) {
    return normalized
  }

  if (normalized.startsWith('geospatial_')) {
    const stripped = normalized.slice('geospatial_'.length)
    if (TOOL_IDENTITIES[stripped]) {
      return stripped
    }
  }

  return normalized
}

export function resolveToolIdentity(toolName: string | null | undefined): ResolvedToolIdentity {
  const rawName = toolName ?? ''
  const name = canonicalToolName(rawName)
  const identity = TOOL_IDENTITIES[name]
  if (identity) {
    return {
      name,
      rawName,
      known: true,
      ...identity,
    }
  }

  return {
    name,
    rawName,
    known: false,
    label: '运行时工具',
    family: 'tool',
    summary: '未在当前应用工具映射中注册的运行时工具。',
    icon: 'wrench',
  }
}

export function resolveToolIdentityForState(toolName: string | null | undefined, state: Record<string, unknown> | null | undefined): ResolvedToolIdentity {
  const identity = resolveToolIdentity(toolName)
  if (identity.name !== 'record_run_evidence') {
    return identity
  }

  return {
    ...identity,
    label: evidenceToolLabel(state) ?? identity.label,
  }
}

function evidenceToolLabel(state: Record<string, unknown> | null | undefined): string | null {
  const input = state?.input
  if (!input || typeof input !== 'object') {
    return null
  }
  const payload = input as Record<string, unknown>
  const recordType = typeof payload.record_type === 'string' ? payload.record_type : ''
  if (recordType === 'artifact') {
    const artifactStage = typeof payload.artifact_stage === 'string' ? payload.artifact_stage : ''
    if (artifactStage === 'final') return '记录最终产物'
    if (artifactStage === 'intermediate') return '记录中间产物'
    return '记录产物'
  }
  return EVIDENCE_RECORD_LABELS[recordType] ?? null
}

File: app/services/tool_identity.py
from __future__ import annotations

from dataclasses import dataclass
from functools import lru_cache
import json
from pathlib import Path


@dataclass(frozen=True, slots=True)
class ToolIdentity:
    name: str
    label: str
    family: str
    summary: str
    icon: str
    known: bool = True


def _repo_root() -> Path:
    return Path(__file__).resolve().parents[2]


def _identity_source_path() -> Path:
    return _repo_root() / 'frontend' / 'src' / 'shared' / 'toolIdentity.json'


@lru_cache(maxsize=1)
def load_tool_identity_map() -> dict[str, ToolIdentity]:
    raw = json.loads(_identity_source_path().read_text(encoding='utf-8'))
    return {
        runtime_name: ToolIdentity(
            name=runtime_name,
            label=str(payload['label']),
            family=str(payload['family']),
            summary=str(payload['summary']),
            icon=str(payload['icon']),
        )
        for runtime_name, payload in raw.items()
    }


def canonical_tool_name(tool_name: str | None) -> str:
    normalized = str(tool_name or '').strip()
    if normalized.startswith('mcp__geospatial__'):
        normalized = normalized.removeprefix('mcp__geospatial__')

    while normalized.startswith('geospatial_geospatial_'):
        normalized = normalized.removeprefix('geospatial_')

    identities = load_tool_identity_map()
    if normalized in identities:
        return normalized

    if normalized.startswith('geospatial_'):
        stripped = normalized.removeprefix('geospatial_')
        if stripped in identities:
            return stripped

    return normalized


def resolve_tool_identity(tool_name: str | None) -> ToolIdentity:
    canonical_name = canonical_tool_name(tool_name)
    identity = load_tool_identity_map().get(canonical_name)
    if identity is not None:
        return identity
    return ToolIdentity(
        name=canonical_name,
        label='运行时工具',
        family='tool',
        summary='未在当前应用工具映射中注册的运行时工具。',
        icon='wrench',
        known=False,
    )
